client/src/main.cpp

```diff
@@ -25,19 +25,40 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 #include "schema/model_generated.h"
 
+#include <Maszyna.Diagnostics/Trace.hxx>
+
 using namespace gl33core;
 
+//  Set this to 'Normal', to compile-out any calls to
+//  MASZYNA_TRACE_WRITELINE with Debug level at compile-time.
+//                                           |
+//                                           V
+MASZYNA_DECLARE_TRACE_SWITCH(AppMain, Debug, Debug);
+MASZYNA_DEFINE_TRACE_SWITCH(AppMain);
+
 const int GLFW_TRUE = static_cast<int>(GL_TRUE);
 const int GLFW_FALSE = static_cast<int>(GL_FALSE);
 
 void error_callback(int, const char* description)
 {
-    std::cerr << description << std::endl;
+    Maszyna::Diagnostics::Trace::Fail("OpenGL error", description);
+}
+
+void key_callback(GLFWwindow* window, int key, int scancode, int action, int modifiers)
+{
+    MASZYNA_TRACE_WRITELINE(AppMain, Debug, "Window: %p, Key: %d, Scancode: %08x, Action: %08x, Modifiers: %08x",
```

Comment:
I guess we can live with cost of extra call, also if we have constexpr for level checking then generated body would be no-op and should be optimized out ;-)